pep-perennial.rst

```diff
@@ -0,0 +1,178 @@
+
+Distributing compiled code for Linux is more complicated than for other popular
+operating systems, because the Linux kernel and the libraries which typically
+accompany it are built in different configurations and combinations for different
+distributions. However, there are certain core libraries which can be expected in
+many common distributions and are reliably backwards compatible, so binaries
+built with older versions of these libraries will work with newer versions.
+:pep:`513` describes these ideas in much more detail.
+
+The ``manylinux1`` (:pep:`513`) and ``manylinux2010`` (:pep:`571`) tags make
+use of these features. They define a set of core libraries and symbol versions
+which wheels may expect on the system, based on CentOS 5 and 6 respectively.
+Typically, packages are built in Docker containers based on these old CentOS
+versions, and then the ``auditwheel`` tool is used to check them and bundle any
+other linked libraries into the wheel.
+
+If we were to define a ``manylinux2014`` tag based on CentOS 7, there would be
+five steps involved to make it practically useful:
+
+1. Write a PEP
+2. Prepare docker images based on CentOS 7.
+3. Add the definition to auditwheel
+4. Allow uploads with the new tag on PyPI
+5. Add code to pip to recognise the new tag and check if the platform is
+   compatible
+
+Although preparing the docker images and updating auditwheel take more work,
+these parts can be used as soon as that work is complete. The changes to pip
+are more straightforward, but not all users will promptly install a new version
+of pip, so distributors are concerned about moving to a new tag too quickly.
+
+This PEP aims to remove the need for steps 1 and 5 above, so new manylinux tags
+can be adopted more easily.
+
+Naming
+======
+
+Tags using the new scheme will look like::
+
+    manylinux_glibc_2_17_x86_64
+
+Where ``2_17`` is the major and minor version of glibc. I.e. for this example,
+the platform must have glibc 2.17 or newer. Installer tools should be prepared
+to handle any numeric values here, but building and publishing wheels to PyPI
+will probably be constrained to specific profiles defined by auditwheel.
+
+The existing manylinux tags can also be represented in the new scheme:
+
+- ``manylinux1_x86_64`` becomes ``manylinux_glibc_2_5_x86_64``
+- ``manylinux2010_x86_64`` becomes ``manylinux_glibc_2_12_x86_64``
+
+``x86_64`` refers to the CPU architecture, as in previous tags.
+
+While this PEP does not attempt to define tags for non-glibc Linux, the name
+glibc is included to leave room for future efforts in that direction.
+
+Wheel compatibility
+===================
+
+There are two components to a tag definition: a specification of what makes a
+compatible wheel, and of what makes a compatible platform.
+
+A wheel may never use symbols from a newer version of glibc than that indicated
+by its tag. Likewise, a wheel with a glibc tag under this scheme may not be
+linked against another libc implementation.
```

Comment:
For the time being, we're only trying to specify binary compatibility for platforms based on glibc, because that's what people have done the work to figure out how to make things work.

This doesn't forbid users from using another libc implementation, but this spec is not meant to cover distributing pre-built binaries to those users. They would have to build extension modules from source until someone has figured out another spec.

The main use case I've heard about for an alternative libc is docker containers based on Alpine Linux, which uses musl. One approach that has been suggested is to standardise alpine-specific tags for that use case.